### PR TITLE
fix(tooltip): focus management with non-interactive elements

### LINKED ISSUE #4232: Regression bug: Vanilla Tooltip focus management
I submitted these PRs (https://github.com/carbon-design-system/carbon/pull/3476 and https://github.com/carbon-design-system/carbon/pull/3148) which corrected a11y bugs on Tooltips in Carbon vanilla and in the latest version of Carbon this behavior is no longer working

## What package(s) are you using?

- [x] `carbon-components`
- [ ] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

 - Tooltips without interactive elements should focus the tooltip on open so that screen reader users have access to the content
- The user should be able to click inside the tooltip without it closing the tooltip

> Is this issue related to a specific component?

Interactive tooltip

> What did you expect to happen? What happened instead? What would you like to
> see changed?

- The tooltip (when there are not interactive elements within it) should receive focus
- The user could select content that was within the tooltip without it closing

> What browser are you working in?

Chrome

> What version of the Carbon Design System are you using?

10.6.4

> What offering/product do you work on? Any pressing ship or release dates we
> should be aware of?

Watson Health Design Pattern & Asset Library. We have a release early Nov

## Steps to reproduce the issue

1. Open the tooltip in the demo below
2. the tooltip should receive focus (there is a white border that displays when it gets focus)

https://codepen.io/elizabethsjudd/pen/oNNNgpM?editors=1000

### PR DESCRIPTION
**Pull request by elizabethsjudd**
Closes #4232 

Tooltips with non-interactive elements as the content need to have focus displayed on the tooltip itself. The user should be able to click within a tooltip without it closing.

#### Changelog

**Changed**
- demo for tooltips to show how a tooltip with non-interactive elements behaves
- updated the styling for the heading which was removed during a bad merge
- updated the focus styles to work on the content element of the tooltip
- moved aria attributes from tooltip to content to prevent duplicate information read by screen readers on non-interactive tooltips and updated others for consistency 

#### Testing / Reviewing
Test all three demos under interactive tooltip for regression behavior. The changed behavior should be as described above on the middle demo.  Tooltips should be passing AVT 1, 2, and 3 after this PR is merged.

### CODE REVIEW

**Comment by elizabethsjudd** [comment]
@dakahn I had to move where the aria attributes were on the tooltip with non-interactive elements so the VO didn't repeat the contents.  Should I update all of the tooltips so they are consistent (it won't break a11y on the other demos)?

File: packages/components/tests/spec/tooltip_spec.js [hunk 1]
```diff
@@ -44,6 +44,18 @@ describe('Test tooltip', function() {
...
     });
 
+    it('Should apply focus to a child element upon opening', function() {
+      element.dispatchEvent(new CustomEvent('click', { bubbles: true }));
+      expect(document.activeElement.closest('.bx--tooltip')).not.toBe(null);
+    });
+
+    // @todo currently not working
+    // it('Should remain open upon click within the tooltip', function() {
+    //   floating.classList.add('bx--tooltip--shown');
```
Comment (asudoh) [inline packages/components/tests/spec/tooltip_spec.js]: @elizabethsjudd Seems that this test case passes in my env. Did you really see it failing?

Comment (elizabethsjudd) [reply packages/components/tests/spec/tooltip_spec.js]: The behavior is not working when I test it in the browser. If it's passing then the test is currently wrong.

Comment (asudoh) [reply packages/components/tests/spec/tooltip_spec.js]: OK would you try fixing the test...? More detailed (manual) test script would also help (is it about not wanting tooltip to be closed when user clicks on tooltip body?). Thanks!

Comment (elizabethsjudd) [reply packages/components/tests/spec/tooltip_spec.js]: yes the behavior we want is that when you click on the tooltip body the tooltip remains open.  I'll try updating the test today.

**Comment by elizabethsjudd** [comment]
@asudoh i've added some more tests and comments to the spec file to explain what's going on in the tests vs what's happening in the browser

steps to re-create:
1. Go to this page: https://deploy-preview-4268--the-carbon-components.netlify.com/?nav=tooltip
2. open the first tooltip in the demo
3. click on the text within the tooltip
4. I want the tooltip to remain open but it's closing and returning focus to the trigger

----

@dakahn thoughts on this: https://github.com/carbon-design-system/carbon/pull/4268#issuecomment-539736798

**Comment by asudoh** [comment]
Thanks @elizabethsjudd for further look! Sent a DM to provide some hits on further debugging.

**Comment by elizabethsjudd** [comment]
@asudoh sent a DM with the results of the debugging work.

**Comment by elizabethsjudd** [comment]
@asudoh thanks for the info in the DM. That fixed the issue I was having and this PR is ready for review.

@dakahn @snidersd do you want to verify my work with screen readers. I tested using VO on Safari and had no issues.

**Comment by dakahn** [comment]
If you can verify with JAWS that is always great. 👍 @elizabethsjudd

File: packages/components/tests/spec/tooltip_spec.js [hunk 2]
```diff
@@ -44,6 +44,35 @@ describe('Test tooltip', function() {
...
+      element.dispatchEvent(new CustomEvent('click', { bubbles: true }));
+
+      const content = floating.querySelector('.bx--tooltip__content');
+      content.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
+      expect(floating.classList.contains('bx--tooltip--shown')).toBe(true);
+    });
+
+    // Known bug: https://github.com/carbon-design-system/carbon/issues/3835
+    // it('Should hide the tooltip upon focusing an external element and focus moves to external element', function() {
+    //   element.dispatchEvent(new CustomEvent('click', { bubbles: true }));
```
Comment (asudoh) [inline packages/components/tests/spec/tooltip_spec.js]: Given we seem to have no way to make this test case work, would we remove this?

Comment (elizabethsjudd) [reply packages/components/tests/spec/tooltip_spec.js]: @asudoh this test does work but it fails because there is the known bug that I linked to in the comments which is why I commented it out. This is still a valid bug with tooltips but I feel like it's beyond the scope of the issue I'm trying to resolve with this PR.

File: packages/components/src/components/tooltip/_tooltip.scss [hunk 3]
```diff
@@ -362,6 +362,7 @@
     word-wrap: break-word;
     color: $inverse-01;
 
+    // @todo this can be depreciated in v11 since focus should always be on the content container not the tooltip
```
Comment (emyarod) [inline packages/components/src/components/tooltip/_tooltip.scss]:
```suggestion
    // @todo this can be deprecated in v11 since focus should always be on the content container not the tooltip
```

File: packages/components/src/components/tooltip/_tooltip.scss [hunk 4]
```diff
@@ -398,6 +399,28 @@
       }
     }
 
+    // Tooltips without interactive elements, need to be sequentially and click focusable so screen reader users
+    // get the tooltip's content and we focus the content container instead of the tooltip root so that
+    // focus-wrap doesn't get triggered
+    .#{$prefix}--tooltip__content[tabindex='0'] {
```
Comment (emyarod) [inline packages/components/src/components/tooltip/_tooltip.scss]: just to confirm I understand correctly: in order to avoid triggering the focus trap on tooltips without interactive elements in the tooltip content, `tabindex="0"` needs to be set on `.bx--tooltip__content` to allow sequential keyboard focus and click focus, is that right?

do we need to add any JS to automatically add/remove `tabindex` from `.bx--tooltip__content` if the tooltip does not contain an interactive element?

Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/_tooltip.scss]: @emyarod You are correct with your understanding. In the original PR to set up tooltips with non-interactive elements (this PR is to fix a bad merge between tooltip PRs that broke this functionality) it was mentioned that it should be up to the developer to include the attribute.  Given now that the tabindex is always included (either 0 or -1), I would not be opposed to adding it to the JS to make it easier for the developers.

**Comment by elizabethsjudd** [comment]
@emyarod ready for re-review

**Review by dakahn — APPROVED** [review]

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 4 above)
Comment (emyarod) [reply packages/components/src/components/tooltip/_tooltip.scss]: I think it would be helpful to add that tabindex switching functionality in our tooltip JS but interested in hearing others' thoughts cc @asudoh

Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/_tooltip.scss]: @emyarod I did add it to the JavaScript and I added tests for it.

File: packages/components/tests/spec/tooltip_spec.js (hunk 2 above)
Comment (asudoh) [reply packages/components/tests/spec/tooltip_spec.js]: Probably we can keep just L60? My point is of reflecting our finding where firing `click` event cannot emulate the change in focus.

File: packages/components/src/components/tooltip/_tooltip.scss [hunk 5]
```diff
@@ -398,6 +399,28 @@
...
+      &:focus {
+        outline: 1px solid $ui-background;
+      }
+    }
+
+    // Tooltips with interactive elements, need to be click focusable but not sequentially focusable so the user
+    // can click within the tooltip and not have it close. Because the element is not actionable it does not need
+    // to have a visible focus indicator (OK'd by IBMa)
+    .#{$prefix}--tooltip__content[tabindex='-1'] {
+      &:focus {
```
Comment (emyarod) [inline packages/components/src/components/tooltip/_tooltip.scss]: probably this doesn't need to be nested?

File: packages/components/src/components/tooltip/_tooltip.scss [hunk 6]
```diff
@@ -398,6 +399,28 @@
       }
     }
 
+    // Tooltips without interactive elements, need to be sequentially and click focusable so screen reader users
+    // get the tooltip's content and we focus the content container instead of the tooltip root so that
+    // focus-wrap doesn't get triggered
+    .#{$prefix}--tooltip__content[tabindex='0'] {
+      // 1 rem === .#{$prefix}--tooltip padding
+      margin: calc(-1rem + 2px);
```
Comment (emyarod) [inline packages/components/src/components/tooltip/_tooltip.scss]: should we use a spacing token in place of `1rem`?

**Review by emyarod — COMMENTED** [review]
the new focus behavior looks good to me, just pending design sign off and figuring out how to deal with the broken test

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 6 above)
Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/_tooltip.scss]: the 1rem was a copy and paste directly from the tooltip padding property. If I change this one, I feel like I should also change that one since they need to stay in sync. Agree?

File: packages/components/tests/spec/tooltip_spec.js (hunk 2 above)
Comment (elizabethsjudd) [reply packages/components/tests/spec/tooltip_spec.js]: I can remove the rest of the test if you'd like. For clarification, that `click` is to open the tooltip. The bug is when i apply the `:focus` to the input which does move the focus but to the wrong element (hence the bug)

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 6 above)
Comment (emyarod) [reply packages/components/src/components/tooltip/_tooltip.scss]: yeah I think that would be fine as long as we kept the comments about why we're making those calculations

**Comment by elizabethsjudd** [comment]
@emyarod ready for re-review

**Review by emyarod — COMMENTED** [review]
we'll need to interpolate the spacing token inside `calc` in Sass for it to transpile down to CSS properly

image

**Comment by elizabethsjudd** [comment]
@emyarod good catch. Thank you! ready for re-review

**Review by aagonzales — APPROVED** [review]
🎉 Awesome! We just had a convo about this in our slack [channel](https://ibm-studios.slack.com/archives/C0M053VPT/p1571683357198300) the other day. LGTM!

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 4 above)
Comment (asudoh) [reply packages/components/src/components/tooltip/_tooltip.scss]: Late on this thread, and need a refresher. This is for interactive tooltip, is it correct? If so, does it really make sense to mark the element sequentially-focusable in DOM given the menu is floating, where we cannot make the focus sequence in DOM match the one user sees visually, and the content node is the only focusable node in the floating menu? Note that we "simulate" sequential focus by tracking focus events.

File: packages/components/src/components/tooltip/tooltip.js [hunk 7]
```diff
@@ -137,6 +137,14 @@ class Tooltip extends mixin(
         throw new Error('Cannot find the target tooltip.');
       }
 
+      const content = tooltip.querySelector(this.options.selectorContent);
+      // Set tabindex to 0 if there are no interactive elements within the tooltip;
+      //  otherwise use -1 so that the tooltip is click focusable only
+      content.setAttribute(
+        'tabindex',
+        content.querySelector(settings.selectorTabbable) ? '-1' : '0'
```
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.js]: Please see: https://github.com/carbon-design-system/carbon/pull/4268#discussion_r339263963

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 4 above)
Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/_tooltip.scss]: @asudoh It is for interactive tooltip and the "simulated" sequential focus did not work for screen reader (SR) users. You have to actually move the focus in to the dialog in order for the SR to announce the content.  When there is not interactive elements within the tooltip to focus on nothing was being read which is why we have the content actually focused.  I've made this change in a previous PR (which was approved and merged). This PR is to fix a bad merge between 2 PRs that overwrote this functionality.

Comment (emyarod) [reply packages/components/src/components/tooltip/_tooltip.scss]: that's true, if there is an element with a higher tabindex before the floating menu in the DOM, the tab order might not match what the user expects after launching the tooltip right?

Comment (asudoh) [reply packages/components/src/components/tooltip/_tooltip.scss]: Sorry for possibly not catching the topic before. What I understand/recall is that the floating menu code puts the focus programmatically. Did you find it differently?

File: packages/components/src/components/tooltip/tooltip.js (hunk 7 above)
Comment (asudoh) [reply packages/components/src/components/tooltip/tooltip.js]: Continuation: https://github.com/carbon-design-system/carbon/pull/4268#discussion_r339821005

File: packages/components/src/components/tooltip/_tooltip.scss (hunk 4 above)
Comment (elizabethsjudd) [reply packages/components/src/components/tooltip/_tooltip.scss]: @asudoh Floating menu does support moving the focus but it needs an element to focus which is why we have the `tabindex="0"` on the element so that we know there is something to focus in the tooltip.  This PR does not change any of the JavaScript logic that moves the focus, just ensures there is something to focus otherwise "nothing" gets focus and screen reader user doesn't get the content.

**Comment by elizabethsjudd** [comment]
@asudoh based on our DM conversation today, I've made updates to the tooltip and tests. This PR is ready for re-review. I did re-test that everything worked for screen reader users locally.

**Review by jeanservaas — APPROVED** [review]

File: packages/components/src/components/tooltip/tooltip.js [hunk 8]
```diff
@@ -137,6 +137,12 @@ class Tooltip extends mixin(
         throw new Error('Cannot find the target tooltip.');
       }
 
+      // Set up focus to be moved to the content on tooltips with no interactive elements
+      const content = tooltip.querySelector(this.options.selectorContent);
+      if (content.querySelector(settings.selectorTabbable) === null) {
+        content.setAttribute('data-floating-menu-primary-focus', '');
```
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.js]: Would we introduce `selectorFocusable`, which basically is a copy of [`selectorTabbable`](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/components/src/globals/js/settings.js#L30-L35) with `:not([tabindex='-1'])` removed? And update [`floating-menu.js` code](https://github.com/carbon-design-system/carbon/blob/v10.7.0/packages/components/src/components/floating-menu/floating-menu.js#L325-L335) to focus on first (programmatically) focusable element if attempts to focus on `[data-floating-menu-primary-focus]` and sequentially focusable element fails?

**Review by asudoh — COMMENTED** [review]
Thank you for the update @elizabethsjudd!

**Comment by elizabethsjudd** [comment]
@asudoh ready for re-review

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @elizabethsjudd!

File: packages/components/src/components/floating-menu/floating-menu.js [hunk 9]
```diff
@@ -325,21 +325,32 @@ class FloatingMenu extends mixin(
...
           focusableNode.focus();
         } else {
           this.element.focus();
           if (__DEV__) {
-            const elementTabindex = this.element.getAttribute('tabindex');
             warning(
-              elementTabindex !== null && parseInt(elementTabindex, 10) > -1,
-              'Floating Menus without interactive elements must include tabindex="0" on the floating element.'
+              focusableNode === null,
+              'Floating Menus must at least a programmatic focusable child by adding tabindex="-1" to the content element.'
```
Comment (emyarod) [inline packages/components/src/components/floating-menu/floating-menu.js]: can we reword this to be a bit clearer?

**Review by emyarod — APPROVED** [review]
thanks for adding the note about focus outline removal on nonactionable elements, just one note about the warning when no focusable node is found

File: packages/components/src/components/floating-menu/floating-menu.js (hunk 9 above)
Comment (elizabethsjudd) [reply packages/components/src/components/floating-menu/floating-menu.js]: @emyarod good catch... sometimes my brain goes faster than my fingers.  I've updated the comment.
